optimistix/_solver/quasi_newton.py

```diff
@@ -115,12 +78,615 @@ def __call__(
+    """
+
+    def _roll_and_set(x, new_vals):
+        # roll the matrix
+        x_shifted = jnp.empty_like(x)
+        x_shifted = x_shifted.at[:-1, :-1].set(x[1:, 1:])
+
+        # sort history buffer
+        last_index = index_start % array.shape[0]
+        sort_idx = (jnp.arange(array.shape[0]) + last_index + 1) % array.shape[0]
+        new_vals = new_vals[sort_idx]
+        # set new history values
+        x_shifted = x_shifted.at[-1, :-1].set(new_vals[:-1])
+        return x_shifted
+
+    def _set_at_index(x, new_vals):
+        return x.at[index_start, :-1].set(new_vals[:-1])
+
+    return jax.lax.cond(
+        index_start < array.shape[0], _set_at_index, _roll_and_set, array, new_vals
+    )
+
+
+def roll_symmetric_matrix(array, new_vals, index_start):
+    """Update a symmetric matrix buffer with a new outer product.
+
+    This function updates a symmetric matrix `array`, which stores inner products such
+    as `S_k^T S_k` in L-BFGS-type algorithms. Depending on `index_start`, it either
+    overwrites an existing entry or rolls the buffer and inserts `new_vals` at the end.
+
+    See Equation (3.2) of:
 
-class _AbstractBFGSDFPUpdate(AbstractQuasiNewtonUpdate):
+        Byrd, R. H., Nocedal, J., & Schnabel, R. B. (1994).
+        "Representations of quasi-Newton matrices and their use in limited memory
+        methods." *Mathematical Programming*, 63(1), 129–156.
+
+    **Arguments**:
+
+    - `array`: A square `(n, n)` JAX array representing the symmetric
+        inner product history matrix (e.g., `S_k^T S_k`).
+
+    - `new_vals`: A JAX array of shape `(n,)` containing new inner
+        product values to insert into the matrix.
+
+    - `index_start`: An integer index indicating where to insert
+        `new_vals`. If `index_start >= array.shape[0]`, the matrix is
+        rolled to discard the oldest entry.
+
+    **Returns**:
+
+    An updated `(n, n)` JAX array with `new_vals` inserted, either
+    in-place or after rolling.
+    """
+
+    def _roll_and_set(x, new_vals):
+        # roll matrix
+        x = jnp.empty_like(x).at[:-1, :-1].set(x[1:, 1:])
+
+        # sort history buffer
+        last_index = index_start % array.shape[0]
+        sort_idx = (jnp.arange(array.shape[0]) + last_index + 1) % array.shape[0]
+        new_vals = new_vals[sort_idx]
+
+        # set last row and col
+        x = x.at[-1, :-1].set(new_vals[:-1])
+        return x.at[:, -1].set(new_vals)
+
+    def _set_at_index(x, new_vals):
+        x = x.at[index_start, :-1].set(new_vals[:-1])
+        return x.at[:, index_start].set(new_vals)
+
+    return jax.lax.cond(
+        index_start < array.shape[0], _set_at_index, _roll_and_set, array, new_vals
+    )
+
+
+def _lbfgs_inverse_hessian_operator_fn(
+    pytree: Y,  # TODO name this better?
+    state: _LBFGSInverseHessianUpdateState,
+) -> Y:
+    """
+    LBFGS descent linear operator.
+
+    """
+    history_len = state.inner_history.shape[0]
```

Comment:
Nit: prefer unpacking over indexing. That is, `[history_len] = state.inner_history.shape` insert. This improves readability by making it clear what is the rank of the provided array, and serves as a runtime assertion that the rank is what you think it is.